copyAsync full=true fails AsyncCatcher.catchOp check

### Describe the bug

I have the following code, which is called from within a command handler:
```
 WorldsProvider provider = Bukkit.getServicesManager().load(WorldsProvider.class);
        provider.levelView().cloneAsync(realWorld,
                builder -> {
                    builder.name(realWorld.getName() + "_" + MyPlugin.instance().getName().toLowerCase());
                    builder.directory(builder.directory().getParent().resolve(builder.name()));
                    builder.key(Key.key(MyPlugin.instance().namespace(), realWorld.key().value() + "_reference"));
                }, fullCopy)
                .exceptionally(err -> {
                    logger.severe("Error creating reference world for world " + realWorld.getName());
                    err.printStackTrace();
                    return null;
                })
                .thenAccept(referenceWorld -> {
                    if (referenceWorld == null) {
                        logger.severe("Could not create reference world for world "
                                + realWorld.getName());
                        future.complete(null);
                        return;
                    }
                    logger.info("For world " + realWorldId + ", created new reference world " + referenceWorld.getName()
                            + " (" + referenceWorld.getUID() + ")");
                    future.complete(referenceWorld.getUID());
                });
```
I get the following traces:
```
[22:29:01 ERROR]: Thread Dimension-Data-IO-Worker-3 failed main thread check: world creation
java.lang.Throwable: null
        at org.spigotmc.AsyncCatcher.catchOp(AsyncCatcher.java:9) ~[folia-1.21.8.jar:1.21.8-5-7a1c5c7]
        at worlds-3.6.0-all.jar/net.thenextlvl.worlds.level.PaperLevel.createAsync(PaperLevel.java:75) ~[worlds-3.6.0-all.jar:?]
        at worlds-3.6.0-all.jar/net.thenextlvl.worlds.view.PaperLevelView.lambda$cloneAsync$13(PaperLevelView.java:356) ~[worlds-3.6.0-all.jar:?]
        at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150) ~[?:?]
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]                     at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1810) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
[22:29:01 ERROR]: [<MyPlugin>] Error creating reference world for world world
[22:29:01 WARN]: java.util.concurrent.CompletionException: java.lang.IllegalStateException: Asynchronous world creation![22:29:01 WARN]:        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
[22:29:01 WARN]:        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
[22:29:01 WARN]:        at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1159)
[22:29:01 WARN]:        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
[22:29:01 WARN]:        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1810)
[22:29:01 WARN]:        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
[22:29:01 WARN]:        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
[22:29:01 WARN]:        at java.base/java.lang.Thread.run(Thread.java:1583)
[22:29:01 WARN]: Caused by: java.lang.IllegalStateException: Asynchronous world creation!
[22:29:01 WARN]:        at org.spigotmc.AsyncCatcher.catchOp(AsyncCatcher.java:10)
[22:29:01 WARN]:        at worlds-3.6.0-all.jar//net.thenextlvl.worlds.level.PaperLevel.createAsync(PaperLevel.java:75)
[22:29:01 WARN]:        at worlds-3.6.0-all.jar//net.thenextlvl.worlds.view.PaperLevelView.lambda$cloneAsync$13(PaperLevelView.java:356)
[22:29:01 WARN]:        at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
[22:29:01 WARN]:        ... 5 more
[22:29:01 ERROR]: [<MyPlugin>] Could not create reference world for world world
[22:29:01 INFO]: An error occurred creating a reference world for minecraft:overworld
```

I've also tested this with just using the command: `world clone minecraft:overworld full true name "world_test" key myplugin:overworld_test`

### To Reproduce

1. Run `world clone minecraft:overworld full true name "world_test" key myplugin:overworld_test`

### Expected behaviour

`copyAsync` internally scheduled tasks should check for being on the appropriate threads

### Screenshots / Videos

_No response_

### Error log (if applicable)

_No response_

### Worlds Version

3.6.0

### Server version

```
> version
[22:58:11 INFO]: Checking version, please wait...
[22:58:11 INFO]: This server is running Folia version 1.21.8-5-ver/1.21.8@7a1c5c7 (2025-09-04T12:50:30Z) (Implementing API version 1.21.8-R0.1-SNAPSHOT)
```

### Checklist

- [x] I am using the newest build from https://github.com/TheNextLvl-net/worlds and the issue still persists.

### Anything else?

_No response_

--- a/src/main/java/net/thenextlvl/worlds/WorldsPlugin.java
+++ b/src/main/java/net/thenextlvl/worlds/WorldsPlugin.java
@@ -35,6 +35,8 @@ import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Locale;
 import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
 import java.util.stream.Stream;
 
 @NullMarked
@@ -150,6 +152,17 @@ public class WorldsPlugin extends JavaPlugin implements WorldsProvider {
                 })
                 .seed(world.getSeed())
                 .name(world.getName());
+    }
+    
+    public <T> CompletableFuture<T> supplyGlobal(Supplier<CompletableFuture<T>> supplier) {
+        var future = new CompletableFuture<T>();
+        getServer().getGlobalRegionScheduler().execute(this, () -> {
+            supplier.get().thenAccept(future::complete).exceptionally(e -> {
+                future.completeExceptionally(e);
+                return null;
+            });
+        });
+        return future;
     }
 
     @Override

--- a/src/main/java/net/thenextlvl/worlds/level/PaperLevel.java
+++ b/src/main/java/net/thenextlvl/worlds/level/PaperLevel.java
@@ -50,7 +50,6 @@ import org.bukkit.craftbukkit.generator.CraftWorldInfo;
 import org.bukkit.event.world.WorldLoadEvent;
 import org.bukkit.generator.WorldInfo;
 import org.jspecify.annotations.NullMarked;
-import org.spigotmc.AsyncCatcher;
 
 import java.io.File;
 import java.io.IOException;
@@ -67,13 +66,15 @@ class PaperLevel extends LevelData {
         super(plugin, builder);
     }
 
+    @Override
+    public CompletableFuture<World> createAsync() {
+        return plugin.supplyGlobal(this::createInternal);
+    }
+
     /**
      * @see CraftServer#createWorld(org.bukkit.WorldCreator)
      */
-    @Override
-    public CompletableFuture<World> createAsync() {
-        AsyncCatcher.catchOp("world creation");
-
+    private CompletableFuture<World> createInternal() {
         var server = ((CraftServer) plugin.getServer());
         var console = server.getServer();
 

--- a/src/main/java/net/thenextlvl/worlds/view/FoliaLevelView.java
+++ b/src/main/java/net/thenextlvl/worlds/view/FoliaLevelView.java
@@ -59,14 +59,9 @@ public class FoliaLevelView extends PaperLevelView {
         if (!handle.players().isEmpty())
             return CompletableFuture.completedFuture(false);
 
-
-        var future = new CompletableFuture<Boolean>();
-
-        plugin.getServer().getGlobalRegionScheduler().execute(plugin, () -> {
-            future.complete(new WorldUnloadEvent(handle.getWorld()).callEvent());
-        });
-
-        return future.thenCompose(success -> {
+        return plugin.supplyGlobal(() -> {
+            return CompletableFuture.completedFuture(new WorldUnloadEvent(handle.getWorld()).callEvent());
+        }).thenCompose(success -> {
             if (!success) return CompletableFuture.completedFuture(false);
 
             var saving = save ? saveAsync(world, true) : CompletableFuture.completedFuture(null);
@@ -76,20 +71,16 @@ public class FoliaLevelView extends PaperLevelView {
                     plugin.getComponentLogger().error("Error during world save", throwable);
                 }
 
-                var closingFuture = new CompletableFuture<@Nullable Void>();
-                plugin.getServer().getRegionScheduler().run(plugin, world, 0, 0, task -> {
+                return plugin.supplyGlobal(() -> {
                     try {
                         handle.getChunkSource().close(false);
                         FeatureHooks.closeEntityManager(handle, save);
                         handle.levelStorageAccess.close();
                     } catch (Exception e) {
                         plugin.getComponentLogger().error("Failed to properly close world after saving", e);
-                    } finally {
-                        closingFuture.complete(null);
                     }
+                    return CompletableFuture.completedFuture(null);
                 });
-
-                return closingFuture;
             }).thenCompose(self -> self).thenApply(ignored -> {
                 try {
                     var field = server.getClass().getDeclaredField("worlds");

--- a/src/main/java/net/thenextlvl/worlds/view/PaperLevelView.java
+++ b/src/main/java/net/thenextlvl/worlds/view/PaperLevelView.java
@@ -28,7 +28,6 @@ import org.bukkit.plugin.java.JavaPlugin;
 import org.jetbrains.annotations.Unmodifiable;
 import org.jspecify.annotations.NullMarked;
 import org.jspecify.annotations.Nullable;
-import org.spigotmc.AsyncCatcher;
 
 import java.io.File;
 import java.io.IOException;
@@ -170,15 +169,13 @@ public class PaperLevelView implements LevelView {
     @Override
     public CompletableFuture<Boolean> unloadAsync(World world, boolean save) {
         return saveLevelDataAsync(world).thenCompose(ignored -> {
-            var future = new CompletableFuture<Boolean>();
-            plugin.getServer().getGlobalRegionScheduler().run(plugin, scheduledTask -> {
+            return plugin.supplyGlobal(() -> {
                 var dragonBattle = world.getEnderDragonBattle();
-                if (plugin.getServer().unloadWorld(world, save)) {
-                    if (dragonBattle != null) dragonBattle.getBossBar().removeAll();
-                    future.complete(true);
-                } else future.complete(false);
+                if (!plugin.getServer().unloadWorld(world, save))
+                    return CompletableFuture.completedFuture(false);
+                if (dragonBattle != null) dragonBattle.getBossBar().removeAll();
+                return CompletableFuture.completedFuture(true);
             });
-            return future;
         }).exceptionally(throwable -> {
             plugin.getComponentLogger().warn("Failed to save level data before unloading", throwable);
             return false;
@@ -190,20 +187,18 @@ public class PaperLevelView implements LevelView {
      */
     @Override
     public CompletableFuture<@Nullable Void> saveAsync(World world, boolean flush) {
-        var future = new CompletableFuture<@Nullable Void>();
-        plugin.getServer().getGlobalRegionScheduler().execute(plugin, () -> {
+        return plugin.supplyGlobal(() -> {
             try {
                 var level = ((CraftWorld) world).getHandle();
                 var oldSave = level.noSave;
                 level.noSave = false;
                 level.save(null, flush, false);
                 level.noSave = oldSave;
-                future.complete(null);
+                return CompletableFuture.completedFuture(null);
             } catch (Exception e) {
-                future.completeExceptionally(e);
+                return CompletableFuture.failedFuture(e);
             }
-        });
-        return future.thenRun(() -> saveLevelDataAsync(world));
+        }).thenRun(() -> saveLevelDataAsync(world));
     }
 
     /**
@@ -236,7 +231,10 @@ public class PaperLevelView implements LevelView {
 
     @Override
     public CompletableFuture<Long> backupAsync(World world) {
-        AsyncCatcher.catchOp("world backup");
+        return plugin.supplyGlobal(() -> backupInternal(world));
+    }
+
+    private CompletableFuture<Long> backupInternal(World world) {
         new WorldBackupEvent(world).callEvent();
         return saveAsync(world, true).thenComposeAsync(ignored -> {
             try {
@@ -326,10 +324,12 @@ public class PaperLevelView implements LevelView {
         return candidate;
     }
 
-
     @Override
     public CompletableFuture<World> cloneAsync(World world, Consumer<Level.Builder> builder, boolean full) {
-        AsyncCatcher.catchOp("world cloning");
+        return plugin.supplyGlobal(() -> cloneInternal(world, builder, full));
+    }
+
+    private CompletableFuture<World> cloneInternal(World world, Consumer<Level.Builder> builder, boolean full) {
         var levelBuilder = plugin.levelBuilder(world);
 
         levelBuilder.name(findFreeName(world.getName()));
@@ -362,8 +362,7 @@ public class PaperLevelView implements LevelView {
 
     @Override
     public CompletableFuture<DeletionResult> deleteAsync(World world, boolean schedule) {
-        AsyncCatcher.catchOp("world deletion");
-        return schedule ? CompletableFuture.completedFuture(scheduleDeletion(world)) : deleteNow(world);
+        return plugin.supplyGlobal(() -> schedule ? CompletableFuture.completedFuture(scheduleDeletion(world)) : deleteNow(world));
     }
 
     @Override
@@ -378,8 +377,7 @@ public class PaperLevelView implements LevelView {
 
     @Override
     public CompletableFuture<DeletionResult> regenerateAsync(World world, boolean schedule) {
-        AsyncCatcher.catchOp("world regeneration");
-        return schedule ? CompletableFuture.completedFuture(scheduleRegeneration(world)) : regenerateNow(world);
+        return plugin.supplyGlobal(() -> schedule ? CompletableFuture.completedFuture(scheduleRegeneration(world)) : regenerateNow(world));
     }
 
     @Override

